service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsController.java

```diff
@@ -0,0 +1,162 @@
+        this.mapper = mapper;
+        this.eventValidationService = eventValidationService;
+        this.docmosisDocumentGeneratorService = docmosisDocumentGeneratorService;
+        this.uploadDocumentService = uploadDocumentService;
+        this.caseDataExtractionService = caseDataExtractionService;
+        this.hearingBookingService = hearingBookingService;
+        this.dateFormatterService = dateFormatterService;
+    }
+
+    @PostMapping("/about-to-start")
+    public AboutToStartOrSubmitCallbackResponse handleMidEvent(@RequestBody CallbackRequest callbackrequest) {
+        CaseDetails caseDetails = callbackrequest.getCaseDetails();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        HearingBooking hearingBooking = hearingBookingService.getMostUrgentHearingBooking(caseData);
+
+        String formattedDate = dateFormatterService.formatLocalDateToString(hearingBooking.getDate() != null
+            ? hearingBooking.getDate() : LocalDate.now(), FormatStyle.LONG);
+
+        caseDetails.getData().put("proceedingLabel", String.format("The case management hearing will be on the %s.",
+            formattedDate));
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(caseDetails.getData())
+            .errors(eventValidationService.validateEvent(caseData, NoticeOfProceedingsGroup.class))
+            .build();
+    }
+
+    @PostMapping("/about-to-submit")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToSubmitEvent(
+            @RequestHeader(value = "authorization") String authorization,
+            @RequestHeader(value = "user-id") String userId,
+            @RequestBody @NotNull CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        Map<String, String> templateData = caseDataExtractionService
+            .getNoticeOfProceedingTemplateData(caseData, caseDetails.getJurisdiction());
+
+        List<DocmosisTemplates> templateTypes = getProceedingTemplateTypes(caseData);
+
+        List<Document> uploadedDocuments = generateAndUploadDocuments(userId, authorization, templateData,
+            templateTypes);
+
+        List<Element> documentsBundle = uploadedDocuments.stream()
+            .map(document -> {
+                return Element.builder()
+                    .id(UUID.randomUUID())
+                    .value(ImmutableMap.builder()
+                        .put("document", DocumentReference.builder()
+                            .document_filename(document.originalDocumentName)
+                            .document_url(document.links.self.href)
+                            .document_binary_url(document.links.binary.href)
+                            .build())
+                        .build())
+                    .build();
+            }).collect(Collectors.toList());
```

Comment:
I think this could be extracted from this controller method. The top level controller should be doing as much as possible. Maybe even additional service would be beneficial but I am not massively fussed.